### PR TITLE
build gstreamer-media-sdk using meson

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
I needed to do that today in order to integrate these plugins nicely within gst-build, I think most of the functionality from cmake is preserved.

### CODE REVIEW

**Comment by ishmael1985** [comment]
Thanks for the meson build script. Indeed we have transitioned to meson for a while, but we also want to merge Windows support with the work. Can you ensure the meson script you wrote is sync-ed with ours so that the effort is minimized when Windows and Linux codebases are merged?

https://github.com/ishmael1985/gstreamer-media-SDK/tree/mfxsink_wip_windows

**Comment by MathieuDuponchelle** [comment]
@ishmael1985 , I wasn't aware of that branch. May I ask what repository should be considered the upstream here ? As far as I can tell, the branch you linked and this repository have diverged pretty significantly.

Regarding the build scripts, I tried to keep the changeset as non-intrusive as possible, and preserved the current cmake build system, it's a different approach from the one that was picked in the branch you linked. I also see that this branch now builds several libraries, whereas this proposal builds a single library as the cmake build system did.

Regarding the meson options, I've made most of these 'combo': ['yes', 'no', 'auto'], in order for the build system to determine by default what exactly can be built.

I was curious as to how you handled the discovery of the media-sdk location, I simply use an option with a default of `/opt/intel/mediasdk`, but I don't lookup the environment variable because there's no native way to do so with meson, I think your approach is more complete and backward compatible. I'm curious about that line though: `mfx_libdir = [mfx_home + '/lib/x64']` , as far as I can tell my version of the media sdk has its lib in lin_x64 , which is where the cmake build system looks it up.

On your branch, a lot of the files have been moved / renamed, part of them in the port to meson commit itself, it makes it more than delicate to "sync our scripts" as is I'm afraid.

As for the other differences, it seems you've got the windows stuff and I've got the Linux stuff (x11 / wayland) :)

I'll have further comments once I know how both repositories stand in relation to the other :)

**Comment by ishmael1985** [comment]
@MathieuDuponchelle, thanks for contributing and taking note of the developments going on gst-mfx plugins. I just referenced my branch so that you are notified of what's going on and what is anticipated in the future. For now 01.org is still mainline for stable development of these plugins, however a lot of cutting-edge development work is being done on my repo that may be merged to mainline once validated as stable and planned for proposal to gst-plugins-bad once the Windows support is complete.
For now the Windows and Linux codebases have diverged significantly but fundamentally share the same architecture and code structure, the gap between the codebases is temporary as soon as we make sure that the Windows version is finally stable on our side, feature-wise it is now complete. Once we iron out the remaining bugs we will unify the codebases in my repo, then propose to both 01.org and GStreamer mainstream for validation and inclusion. We will finally have a fully hardware-accelerated set of MSDK plugins that works on both platforms under the same codebase, I believe this goal alone is worth pursuing this development :)
Since we are finally working on this, we might as well fully transition to Meson from CMake, so while your Meson script is good for the current version, I just want you to avoid a lot of effort re-writing the Meson script in case you plan to take up the unified version, so really it is just a notification, but seriously, good work on the script.
The one who wrote the initial version of the meson script for the Windows version is from this guy (https://github.com/ph0b), so honestly haven't studied it too much yet, and I'm no meson expert. Normally for these cases I would try to detect the MFX_HOME env variable and if I can't find it I would set it to default /opt/intel/mediasdk path. You also need to note that the installation paths are a little different for both Windows and Linux versions, if possible, try to account for those differences in your script, a little bit of platform auto-detection would help :)
If you're interested in helping me out on the effort of unifying the codebases please feel free to contribute to my branch. Reinforcements are always welcome :)

**Comment by ph0b** [comment]
@MathieuDuponchelle good job on the configuration, thanks for contributing it!

It will not be complicated to merge my part back in later as you're correct: I've got the windows part and you've got the (bigger) linux part. 

So in my opinion it's fine to merge your part here on this repo. I can add the MFX_HOME var detection and update the windows side to reflect your changes. 

One thing though: I'd rather use booleans rather than yes/no/auto. The auto use case doesn't cover enough cases to be relevant in my view, while booleans are cleaner/clearer to use.

**Comment by MathieuDuponchelle** [comment]
Hey @ph0b , thanks for the positive reaction :) Regarding the options, I do have a few using booleans, however for cases such as `MFX_H264_ENCODER` I'd rather have three-valued logic, otherwise I think it's a bit confusing what should happen when you have `MFX_ENCODER` set to False and `MFX_H264_ENCODER` or such set to True, but I'm not hell-bent on that.

For the X-related variables, with soft dependencies, I'm pretty positive three-valued logic is appropriate however, don't you agree ?

**Comment by MathieuDuponchelle** [comment]
@ishmael1985 , thanks for the clarification by the way :) I must admit working on the Windows side of things is out of my scope, but I definitely don't want to make it awkward to merge it later on.

**Comment by MathieuDuponchelle** [comment]
@ishmael1985 , so what's the plan for this pull request ? Can it be merged here ?

**Comment by SiewHoon** [comment]
Hi Mathieu,

Tomorrow I will do it. Can pls submit issue at here? And include issue number inside your patch?  I will check on it and run some quick test on my Linux machine before merge in.

Sorry for delay.

Thank you.

Sent from my iPhone

On 12 Jun 2017, at 20:47, Mathieu Duponchelle <notifications@github.com<mailto:notifications@github.com>> wrote:


@ishmael1985<https://github.com/ishmael1985> , so what's the plan for this pull request ? Can it be merged here ?

**Comment by ishmael1985** [comment]
@MathieuDuponchelle, once it gets merged, the README has to be updated as well...

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,205 @@
+# This project use C source code
+project('GstMfx', 'c',
+	version: '1.3.2',
+	meson_version: '>= 0.35.0',
+  	default_options : ['warning_level=1',
```
Comment (SiewHoon) [inline meson.build]: Remove white space.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+			endif
+		endforeach
+	endif
+else
+	foreach opt : ['WITH_WAYLAND', 'WITH_X11', 'USE_EGL_RENDERER']
+		if get_option (opt) == 'yes'
+			error('@0@ required but MFX_SINK is false'.format(opt))
+		endif
+	endforeach
+endif
```
Comment (SiewHoon) [inline meson.build]: Remove the white space.

**Comment by SiewHoon** [comment]
@MathieuDuponchelle, linux version meson very nice to use. But sorry, I still have to request to fix the whitespace issue and re-submit the patch before I can merge in.  Thank you.

**Comment by MathieuDuponchelle** [comment]
@SiewHoon sure, fixed

**Comment by ishmael1985** [comment]
@MathieuDuponchelle, you can check out the unified version which works on both Windows and Linux (no meson script here though), so you may consider to try to re-work your script with the new codebase.

https://github.com/ishmael1985/gstreamer-media-SDK/tree/mfxsink_wip_windows

**Comment by ishmael1985** [comment]
@MathieuDuponchelle, I have refactored your script more or less to include Windows compilation support, plus some nice touches I think, please tell me what you think of it. Thanks :)

https://github.com/ishmael1985/gstreamer-media-SDK/tree/mfxsink_wip_windows

**Pull request closed without merging**